### PR TITLE
Optional string concatenation 

### PR DESCRIPTION
**Pull request by illusional**
From: #5464 
Closes: #5460 

Recreated to be in the Cromwell repo with access to testing. Thanks @aednichols!

### CODE REVIEW

**Comment by illusional** [comment]
Looks like there is a test that's guarding against my change. I believe this test is incorrect, but I'll leave that confirmation to you (maybe @cjllanwarne as it looks like you wrote the original test):

[`ValueEvaluatorSpec.scala#L538-L542`](https://github.com/broadinstitute/cromwell/blob/50f28da6a665526c1bdb1d5528400ee9deeaa5d4/wdl/model/draft2/src/test/scala/wdl/expression/ValueEvaluatorSpec.scala#L538-L542)
```scala
  "Optional values" should "fail to perform addition with the + operator if the argument is None" in {
    val hello = WomString("hello ")
    val noneWorld = WomOptionalValue.none(WomStringType)
    hello.add(noneWorld) should be(Failure(OptionalNotSuppliedException("+")))
  }
```

Relevant text from original PR:

The [WDL Spec: _Interpolating and concatenating optional strings_](https://github.com/openwdl/wdl/blob/master/versions/development/SPEC.md#interpolating-and-concatenating-optional-strings)

> Within interpolations, string concatenation with the + operator has special typing properties to facilitate formulation of command-line flags. [...] If either operand has an optional type, then the concatenation has type String?, and the runtime result is None if either operand is None

**Comment by cjllanwarne** [comment]
@illusional it looks like you're changing the global behavior of `+` with respect to optional strings, whereas the spec is asking specifically within interpolations.

That discrepancy between "ok in interpolations" and "not ok outside interpolations" is usually handled by catching the `OptionalNotSuppliedException` in the interpolation logic and replacing the entire output with an empty string. I don't know for sure why that's not working in this case but that would be my first suggestion on where to look

**Comment by cjllanwarne** [comment]
@illusional I'm not sure that it'll work, but my first guess would be - can we replace your changes with returning an empty optional value here, instead of returning an empty string: https://github.com/broadinstitute/cromwell/blob/develop/wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/values/BinaryOperatorEvaluators.scala#L52

**Comment by illusional** [comment]
Hey @cjllanwarne, I think that's maybe what this PR should be. I'll have a look at testing that today.

**Comment by illusional** [comment]
Hey @cjllanwarne, thanks for the suggestion. I've changed it to return an Optional string value and it seems to pass my test. I've force pushed this branch to the new set of changes to re-run the tests.

File: centaur/src/main/resources/standardTestCases/string_interpolation_optional/string_interpolation_optional.wdl [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    String? str
+  }
+
+  command <<<
+      echo ~{'"' + str + '"'}
+  >>>
+
+  output {
+    String out = read_string(stdout())
+  }
```
Comment (mcovarr) [inline centaur/src/main/resources/standardTestCases/string_interpolation_optional/string_interpolation_optional.wdl:14]: It looks like a missing `docker` runtime attribute is responsible for the non-`Local` Centaur failures
```suggestion
  }
  
  runtime {
    docker: "ubuntu:latest"
  }
```

File: centaur/src/main/resources/standardTestCases/string_interpolation_optional/string_interpolation_optional.wdl [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+}
+
+workflow echo_wf {
```
Comment (mcovarr) [inline centaur/src/main/resources/standardTestCases/string_interpolation_optional/string_interpolation_optional.wdl:17]: (Stops my IntelliJ / Winstanley from grouching)
```suggestion
workflow echo_wf {
  input {
    String? str
  }
```

File: centaur/src/main/resources/standardTestCases/string_interpolation_optional/string_interpolation_optional.wdl [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  call string_interpolation_task
```
Comment (mcovarr) [inline centaur/src/main/resources/standardTestCases/string_interpolation_optional/string_interpolation_optional.wdl:18]: IntelliJ / Winstanley degrouching part 2
```suggestion
  call string_interpolation_task { input: str = str }
```

**Comment by cjllanwarne** [comment]
Looks good to me too. Thanks @illusional!

**Comment by illusional** [comment]
Thanks @mcovarr for the code suggestion fixes, and thanks all for merging :)

Sorry I was a bit uncommunicative on this PR, but thanks for merging it!

**Pull request merged**